src/upload/index.ts

```diff
@@ -0,0 +1,290 @@
+import {
+  GetObjectCommand,
+  HeadObjectCommand,
+  NoSuchKey,
+  NotFound,
+  PutObjectCommand,
+  S3Client,
+} from "@aws-sdk/client-s3";
+import zlib from "zlib";
+import fs from "fs";
+
+import { LightBlockCache, lightBlockCache } from "@/cache";
+import { LightBlock } from "@/models/lightstreamer";
+import { logger } from "@/utils/logger";
+
+class UploadError extends Error {}
+
+export type BlockRange = {
+  start: number;
+  end: number;
+};
+
+export type BlockFile = {
+  blocks: string;
+  manifest: string;
+  timestamp: number;
+};
+
+export class LightBlockUpload {
+  private cache: LightBlockCache;
+  private s3Client: S3Client;
+  private chunkSizeMb: number;
+  private bucket: string;
+  private latestPath = "latest.json";
+  private blockFileName = "blocks";
+
+  constructor(cache: LightBlockCache) {
+    this.cache = cache;
+    if (!process.env["BUCKET_ENDPOINT"]) {
+      throw new UploadError("BUCKET_ENDPOINT not set");
+    }
+    if (!process.env["BUCKET_ACCESS_KEY_ID"]) {
+      throw new UploadError("BUCKET_ACCESS_KEY_ID not set");
+    }
+    if (!process.env["BUCKET_SECRET_ACCESS_KEY"]) {
+      throw new UploadError("BUCKET_SECRET_ACCESS_KEY not set");
+    }
+    if (!process.env["BUCKET_NAME"]) {
+      throw new UploadError("BUCKET_NAME not set");
+    }
+    if (!process.env["UPLOAD_CHUNK_SIZE_MB"]) {
+      throw new UploadError("UPLOAD_CHUNK_SIZE_MB not set");
+    }
+
+    this.chunkSizeMb = parseInt(
+      process.env["UPLOAD_CHUNK_SIZE_MB"] as string,
+      10,
+    );
+    this.bucket = process.env["BUCKET_NAME"];
+    this.s3Client = new S3Client({
+      region: "auto",
+      endpoint: process.env["BUCKET_ENDPOINT"],
+      credentials: {
+        accessKeyId: process.env["BUCKET_ACCESS_KEY_ID"],
+        secretAccessKey: process.env["BUCKET_SECRET_ACCESS_KEY"],
+      },
+    });
+  }
+
+  async upload(): Promise<void> {
+    const latestJson = await this.getObject(this.latestPath);
+
+    let currentUploadSize = 0;
+    if (!latestJson) {
+      console.warn("No latest json, starting upload from beginning...");
+    } else {
+      const latest: BlockFile = JSON.parse(latestJson);
+      currentUploadSize = (await this.getFileSize(latest.blocks)) || 0;
+    }
+
+    try {
+      logger.info(
+        `Current file uploaded size: ${this.bytesToMbRounded(
+          currentUploadSize,
+        )} MB, creating new upload...`,
+      );
+      const files = await this.createBlockFiles(
+        this.blockFileName,
+        currentUploadSize,
+      );
+      const prefix = String(files.timestamp) + "/";
```

Comment:
What is going to purge the old block files, and prevent them from endlessly accumulating? IIRC, AWS S3 has bucket policies that allow automatic deletion of objects, does R2 have anything similar? Is the plan to use that?

Reply:
Yea I have already set the retention on these buckets to 3 days. I calculated that at our current data accumulation rate we are getting about 1MB per day. So I set the chunk size to 1MB, if our blocks get bigger it is possible we start accumulating much more, so I can set an alert on this if our bucket grows too fast, we can modify the chunk size or scale back the retention.

I just didn't want to write a bunch of new logic for cleanup in this service if we don't have to.